pkg/code-generator/schemagen/cue.go

```diff
@@ -0,0 +1,188 @@
+		absoluteRoot:     absoluteRoot,
+		protoDir:         anyvendor.DefaultDepDir,
+	}
+}
+
+func (c *cueGenerator) GetJsonSchemaForProject(project *model.Project) (map[schema.GroupVersionKind]*apiextv1beta1.JSONSchemaProps, error) {
+	protobufExtractor, err := c.getProtobufExtractorForProject(project)
+	if err != nil {
+		return nil, err
+	}
+
+	instances, err := protobufExtractor.Instances()
+	if err != nil {
+		return nil, err
+	}
+
+	// Convert cuelang to openapi
+	openApiGenerator := &openapi.Generator{
+		// k8s structural schemas do not allow $refs, i.e. all references must be expanded
+		ExpandReferences: true,
+	}
+
+	openApiSchemas := OpenApiSchemas{}
+	built := cue.Build(instances)
+	for _, builtInstance := range built {
+		// Avoid generating openapi for irrelevant proto imports.
+		if !strings.HasSuffix(builtInstance.ImportPath, project.ProjectConfig.GoPackage) {
+			continue
+		}
+
+		if builtInstance.Err != nil {
+			return nil, err
+		}
+		if err = builtInstance.Value().Validate(); err != nil {
+			return nil, eris.Errorf("Cue instance validation failed for %s: %+v", project.ProtoPackage, err)
+		}
+		schemas, err := openApiGenerator.Schemas(builtInstance)
+		if err != nil {
+			return nil, eris.Errorf("Cue openapi generation failed for %s: %+v", project.ProtoPackage, err)
+		}
+
+		// Iterate openapi objects to construct mapping from proto message name to openapi schema
+		for _, kv := range schemas.Pairs() {
+			openApiSchemas[kv.Key] = kv.Value.(*openapi.OrderedMap)
+		}
+
+		return c.convertOpenApiSchemasToJsonSchemas(project, openApiSchemas)
+	}
+
+	return nil, nil
+}
+
+func (c *cueGenerator) getProtobufExtractorForProject(project *model.Project) (*protobuf.Extractor, error) {
+	// Collect all protobuf definitions including transitive dependencies.
+	var imports []string
+	protoRoot := filepath.Join(c.absoluteRoot, c.protoDir)
+
+	for _, projectProto := range project.ProjectConfig.ProjectProtos {
+		importsForResource, err := c.importsCollector.CollectImportsForFile(protoRoot, filepath.Join(protoRoot, projectProto))
+		if err != nil {
+			return nil, err
+		}
+		imports = append(imports, importsForResource...)
+	}
+	imports = stringutils.Unique(imports)
+
+	// Parse protobuf into cuelang
+	protobufExtractor := protobuf.NewExtractor(&protobuf.Config{
+		Root:   protoRoot,
+		Module: project.ProjectConfig.GoPackage,
+		Paths:  imports,
+	})
+
+	for _, projectProto := range project.ProjectConfig.ProjectProtos {
+		if err := protobufExtractor.AddFile(projectProto, nil); err != nil {
+			return nil, err
+		}
+	}
+	return protobufExtractor, nil
+}
+
+func (c *cueGenerator) convertOpenApiSchemasToJsonSchemas(project *model.Project, schemas OpenApiSchemas) (map[schema.GroupVersionKind]*apiextv1beta1.JSONSchemaProps, error) {
+	jsonSchemasByGVK := make(map[schema.GroupVersionKind]*apiextv1beta1.JSONSchemaProps)
+
+	for schemaKey, schemaValue := range schemas {
+		schemaGVK := c.getGVKForSchemaKey(project, schemaKey)
```

Comment:
Can getGVKForSchemaKey ever error here? It looks like it just parses the schemaKey, but it also gets the group and version?

Reply:
It shouldn't. If it does error, it's because the project has been misconfigured